dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/WorkerGroupTaskDispatchManager.java

```diff
@@ -0,0 +1,117 @@
+
+import org.apache.dolphinscheduler.server.master.engine.task.client.ITaskExecutorClient;
+import org.apache.dolphinscheduler.server.master.engine.task.runnable.ITaskExecutionRunnable;
+import org.apache.dolphinscheduler.server.master.runner.queue.DelayEntry;
+import org.apache.dolphinscheduler.server.master.runner.queue.PriorityDelayQueue;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+@Slf4j
+public class WorkerGroupTaskDispatchManager implements AutoCloseable {
+
+    @Autowired
+    private ITaskExecutorClient taskExecutorClient;
+
+    @Getter
+    private final ConcurrentHashMap<String, WorkerGroupTaskDispatchWaitingQueueLooper> workerGroupTaskDispatchWaitingQueueLooperMap;
+    @Getter
+    private final ConcurrentHashMap<String, PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>>> workerGroupPriorityDelayQueueMap;
+
+    public WorkerGroupTaskDispatchManager() {
+        workerGroupTaskDispatchWaitingQueueLooperMap = new ConcurrentHashMap<>();
+        workerGroupPriorityDelayQueueMap = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Adds a task to the specified worker group queue and starts or wakes up the corresponding processing loop.
+     *
+     * @param workerGroup the identifier for the worker group, used to distinguish different task queues
+     * @param taskExecutionRunnable an instance of ITaskExecutionRunnable representing the task to be executed
+     * @param delayTimeMills the delay time before the task is executed, in milliseconds
+     */
+    public void add(String workerGroup, ITaskExecutionRunnable taskExecutionRunnable, long delayTimeMills) {
+        PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>> workerGroupQueue =
+                workerGroupPriorityDelayQueueMap.get(workerGroup);
+        if (workerGroupQueue != null) {
+            workerGroupQueue.add(new DelayEntry<>(delayTimeMills, taskExecutionRunnable));
+            log.info("queue size {}", workerGroupQueue.size());
+        } else {
+            log.error("workerGroup {} not found", workerGroup);
+        }
+    }
+
+    /**
+     * Stops a specific worker group's task dispatch waiting queue looper.
+     *
+     * @param workerGroup the identifier for the worker group
+     */
+    public synchronized void stopWorkerGroup(String workerGroup) throws Exception {
+        WorkerGroupTaskDispatchWaitingQueueLooper looper =
+                workerGroupTaskDispatchWaitingQueueLooperMap.remove(workerGroup);
+        PriorityDelayQueue<DelayEntry<ITaskExecutionRunnable>> workerGroupQueue =
+                workerGroupPriorityDelayQueueMap.get(workerGroup);
+        if (workerGroupQueue != null) {
+            workerGroupQueue.clear();
+        }
+        if (looper != null) {
+            looper.close();
+        }
+    }
```

Comment:
There will exist a problem, you cannot clear the task, the running task which put into the queue should continue running.

Reply:
This method is a workergroup deletion logic, which may not be able to handle even if it is placed back in the global queue